File: backend/src/jobs/queue.ts
import type { JobStatus, JobType, PipelineJobPayload } from "../domain/processing.js";
import type { AppSupabaseClient } from "../lib/supabase.js";

export interface JobRecord {
  id: string;
  projectId: string;
  userId: string;
  jobType: JobType;
  status: JobStatus;
  payload: PipelineJobPayload;
  attempts: number;
  maxAttempts: number;
}

export interface JobRunner {
  runPipeline(payload: PipelineJobPayload, jobId?: string): Promise<void>;
}

export class JobQueue {
  constructor(
    private readonly admin: AppSupabaseClient,
    private readonly runner: JobRunner,
  ) {}

  async enqueuePipeline(payload: PipelineJobPayload): Promise<{ jobId: string; inline: boolean }> {
    const idempotencyKey = `pipeline:${payload.projectId}`;

    const { data: activeJob } = await this.admin
      .from("processing_jobs")
      .select("id, status")
      .eq("idempotency_key", idempotencyKey)
      .in("status", ["queued", "running"])
      .maybeSingle();

    if (activeJob) {
      return { jobId: activeJob.id, inline: false };
    }

    const inserted = await this.tryInsertJob({
      projectId: payload.projectId,
      userId: payload.userId,
      jobType: "project_pipeline",
      payload,
      idempotencyKey,
    });

    if (inserted) {
      void this.runner.runPipeline(payload, inserted.id);
      return { jobId: inserted.id, inline: true };
    }

    void this.runner.runPipeline(payload);
    return { jobId: "inline", inline: true };
  }

  async claimNextJob(): Promise<JobRecord | null> {
    const { data, error } = await this.admin
      .from("processing_jobs")
      .select("*")
      .eq("status", "queued")
      .order("created_at", { ascending: true })
      .limit(1)
      .maybeSingle();

    if (error || !data) return null;

    const { data: claimed, error: claimError } = await this.admin
      .from("processing_jobs")
      .update({
        status: "running",
        started_at: new Date().toISOString(),
        attempts: data.attempts + 1,
        updated_at: new Date().toISOString(),
      })
      .eq("id", data.id)
      .eq("status", "queued")
      .select("*")
      .maybeSingle();

    if (claimError || !claimed) return null;
    return mapJobRecord(claimed);
  }

  async completeJob(jobId: string) {
    await this.admin
      .from("processing_jobs")
      .update({
        status: "completed",
        completed_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      })
      .eq("id", jobId);
  }

  async failJob(jobId: string, message: string) {
    await this.admin
      .from("processing_jobs")
      .update({
        status: "failed",
        error_message: message,
        completed_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      })
      .eq("id", jobId);
  }

  private async tryInsertJob(input: {
    projectId: string;
    userId: string;
    jobType: JobType;
    payload: PipelineJobPayload;
    idempotencyKey: string;
  }) {
    const { data, error } = await this.admin
      .from("processing_jobs")
      .insert({
        project_id: input.projectId,
        user_id: input.userId,
        job_type: input.jobType,
        status: "queued",
        payload: input.payload,
        current_step: "queued",
        idempotency_key: input.idempotencyKey,
      })
      .select("*")
      .single();

    if (error) {
      console.warn("[jobs] DB queue unavailable, running inline:", error.message);
      return null;
    }

    return mapJobRecord(data);
  }
}

function mapJobRecord(row: Record<string, unknown>): JobRecord {
  return {
    id: String(row.id),
    projectId: String(row.project_id),
    userId: String(row.user_id),
    jobType: row.job_type as JobType,
    status: row.status as JobStatus,
    payload: row.payload as PipelineJobPayload,
    attempts: Number(row.attempts ?? 0),
    maxAttempts: Number(row.max_attempts ?? 3),
  };
}

export function startJobPoller(queue: JobQueue, runner: JobRunner, intervalMs = 5000) {
  const timer = setInterval(async () => {
    const job = await queue.claimNextJob();
    if (!job) return;

    try {
      if (job.jobType === "project_pipeline") {
        await runner.runPipeline(job.payload, job.id);
      }
      await queue.completeJob(job.id);
    } catch (error) {
      const message = error instanceof Error ? error.message : "Job failed";
      await queue.failJob(job.id, message);
    }
  }, intervalMs);

  timer.unref?.();
  return () => clearInterval(timer);
}

File: backend/src/captionovo/jobs/queue.py
import asyncio
from dataclasses import dataclass
from datetime import UTC, datetime

from captionovo.domain.processing import PipelineJobPayload


@dataclass
class JobRecord:
    id: str
    project_id: str
    user_id: str
    job_type: str
    status: str
    payload: PipelineJobPayload
    attempts: int
    max_attempts: int


def _map_job_record(row: dict) -> JobRecord:
    payload_data = row["payload"]
    payload = PipelineJobPayload(
        project_id=payload_data["projectId"],
        user_id=payload_data["userId"],
        outputs=payload_data["outputs"],
        transcript_mode=payload_data["transcriptMode"],
        language=payload_data["language"],
        media_type=payload_data["mediaType"],
        storage_path=payload_data["storagePath"],
    )
    return JobRecord(
        id=str(row["id"]),
        project_id=str(row["project_id"]),
        user_id=str(row["user_id"]),
        job_type=row["job_type"],
        status=row["status"],
        payload=payload,
        attempts=int(row.get("attempts") or 0),
        max_attempts=int(row.get("max_attempts") or 3),
    )


def _payload_to_db(payload: PipelineJobPayload) -> dict:
    return {
        "projectId": payload.project_id,
        "userId": payload.user_id,
        "outputs": payload.outputs,
        "transcriptMode": payload.transcript_mode,
        "language": payload.language,
        "mediaType": payload.media_type,
        "storagePath": payload.storage_path,
    }


class JobQueue:
    def __init__(self, admin, runner) -> None:
        self._admin = admin
        self._runner = runner

    async def enqueue_pipeline(self, payload: PipelineJobPayload) -> dict:
        idempotency_key = f"pipeline:{payload.project_id}"

        active = (
            self._admin.table("processing_jobs")
            .select("id, status")
            .eq("idempotency_key", idempotency_key)
            .in_("status", ["queued", "running"])
            .maybe_single()
            .execute()
        )
        if active.data:
            return {"job_id": active.data["id"], "inline": False}

        inserted = self._try_insert_job(
            project_id=payload.project_id,
            user_id=payload.user_id,
            job_type="project_pipeline",
            payload=payload,
            idempotency_key=idempotency_key,
        )

        if inserted:
            asyncio.create_task(self._runner.run_pipeline(payload, inserted.id))
            return {"job_id": inserted.id, "inline": True}

        asyncio.create_task(self._runner.run_pipeline(payload))
        return {"job_id": "inline", "inline": True}

    def claim_next_job(self) -> JobRecord | None:
        result = (
            self._admin.table("processing_jobs")
            .select("*")
            .eq("status", "queued")
            .order("created_at")
            .limit(1)
            .maybe_single()
            .execute()
        )
        if not result.data:
            return None

        data = result.data
        claimed = (
            self._admin.table("processing_jobs")
            .update(
                {
                    "status": "running",
                    "started_at": datetime.now(UTC).isoformat(),
                    "attempts": data.get("attempts", 0) + 1,
                    "updated_at": datetime.now(UTC).isoformat(),
                }
            )
            .eq("id", data["id"])
            .eq("status", "queued")
            .select("*")
            .maybe_single()
            .execute()
        )
        if not claimed.data:
            return None
        return _map_job_record(claimed.data)

    def complete_job(self, job_id: str) -> None:
        self._admin.table("processing_jobs").update(
            {
                "status": "completed",
                "completed_at": datetime.now(UTC).isoformat(),
                "updated_at": datetime.now(UTC).isoformat(),
            }
        ).eq("id", job_id).execute()

    def fail_job(self, job_id: str, message: str) -> None:
        self._admin.table("processing_jobs").update(
            {
                "status": "failed",
                "error_message": message,
                "completed_at": datetime.now(UTC).isoformat(),
                "updated_at": datetime.now(UTC).isoformat(),
            }
        ).eq("id", job_id).execute()

    def _try_insert_job(
        self,
        *,
        project_id: str,
        user_id: str,
        job_type: str,
        payload: PipelineJobPayload,
        idempotency_key: str,
    ) -> JobRecord | None:
        try:
            result = (
                self._admin.table("processing_jobs")
                .insert(
                    {
                        "project_id": project_id,
                        "user_id": user_id,
                        "job_type": job_type,
                        "status": "queued",
                        "payload": _payload_to_db(payload),
                        "current_step": "queued",
                        "idempotency_key": idempotency_key,
                    }
                )
                .select("*")
                .single()
                .execute()
            )
            if not result.data:
                return None
            return _map_job_record(result.data)
        except Exception as exc:
            print(f"[jobs] DB queue unavailable, running inline: {exc}")
            return None


async def start_job_poller(queue: JobQueue, runner, interval_ms: int = 5000):
    while True:
        job = queue.claim_next_job()
        if job:
            try:
                if job.job_type == "project_pipeline":
                    await runner.run_pipeline(job.payload, job.id)
                queue.complete_job(job.id)
            except Exception as exc:
                message = str(exc) if str(exc) else "Job failed"
                queue.fail_job(job.id, message)
        await asyncio.sleep(interval_ms / 1000)
